source/SVInt.cpp

```diff
@@ -86,146 +108,64 @@ SVInt::SVInt(StringRef str) {
-    else
-        pVal = new uint64_t[getNumWords()]();
-
-    SVInt digit(bitWidth, 0, false);
-    SVInt radixSv(bitWidth, radix, false);
-
-    for (; c != str.end(); c++) {
+    // convert the remaining chars to an array of digits to pass to the other
+    // constructor
+    SmallVectorSized<logic_t, 16> digits;
+    for (; c != str.end(); ++c) {
         char d = *c;
         if (d == '_')
             continue;
 
-        int value;
-        int unknown;
+        logic_t value;
         switch (d) {
             case 'X':
             case 'x':
-                value = 0;
-                unknown = ones;
+                value = logic_t::x;
                 break;
             case 'Z':
             case 'z':
             case '?':
-                value = ones;
-                unknown = ones;
+                value = logic_t::z;
                 break;
             default:
-                value = getHexDigitValue(d);
-                unknown = 0;
-                ASSERT(value < radix, "Invalid character");
+                value = logic_t(getHexDigitValue(d));
                 break;
         }
+        digits.append(value);
+    }
 
-        // shift if we can, multiply if we must, making sure we don't
-        // lose the unknownFlag
-        bool myUnknownFlag = unknownFlag;
-        if (shift)
-            *this = shl(shift, false);
-        else
-            *this *= radixSv;
-        if (unknownFlag != myUnknownFlag) {
-            // this means that we just did the first shift or multiply, which
-            // made us lose our unknown flag, and also means that we now think
-            // we are a single word. Let's rectify that
-            unknownFlag = myUnknownFlag;
-            pVal = new uint64_t[getNumWords()]();
-        }
-
-        if (unknownFlag) {
-            // In base ten we can't have individual bits be X or Z, it's all or nothing
-            if (radix == 10) {
-                if (value)
-                    setAllZ();
-                else
-                    setAllX();
-                break;
-            }
-
-            // otherwise, make sure the unknown flag is set for this group of bits
-            // because we're shifting bits for the radix involved (2, 8, or 16) we
-            // know that the bits we're setting are fresh and all zero, so adding
-            // won't cause any kind of carry
-            pVal[0] += value;
-            pVal[getNumWords(bitWidth, false)] += unknown;
-        }
-        else {
-            // add in the new digit
-            if (digit.isSingleWord())
-                digit.val = value;
-            else
-                digit.pVal[0] = value;
-            *this += digit;
-        }
+    if (!isSingleWord()) {
+        // We use the assignment operator to return the result of the other constructor,
+        // but the assignment operator will try to delete our pVal, so let's make sure
+        // there is something there
+        pVal = new uint64_t[0]();
```

Comment:
Deleting a null pointer is fine, there's no reason to waste time allocating here.